pilopt/src/lib.rs

```diff
@@ -552,29 +561,92 @@ fn remove_trivial_identities<T: FieldElement>(pil_file: &mut Analyzed<T>) {
                 }
                 None
             }
-            IdentityKind::Plookup => {
-                assert_eq!(
-                    identity.left.expressions.len(),
-                    identity.right.expressions.len()
-                );
-                identity.left.expressions.is_empty().then_some(index)
+            Identity::Lookup(LookupIdentity { left, right, .. }) => {
+                assert_eq!(left.expressions.len(), right.expressions.len());
+                left.expressions.is_empty().then_some(index)
             }
-            IdentityKind::Permutation => None,
-            IdentityKind::Connect => None,
+            Identity::Permutation(..) => None,
+            Identity::Connect(..) => None,
         })
         .collect();
     pil_file.remove_identities(&to_remove);
 }
 
 fn remove_duplicate_identities<T: FieldElement>(pil_file: &mut Analyzed<T>) {
+    // TODO: there must be a less verbose way to do this...
+    /// Wrapper around `Identity` that implements `PartialEq` and `Ord` for canonical comparison, ignoring source information and id.
+    struct CanonicalIdentity<'a, T>(&'a Identity<T>);
+
+    impl<T: FieldElement> Ord for CanonicalIdentity<'_, T> {
+        fn cmp(&self, other: &Self) -> Ordering {
+            // we implement our own discriminant since std::mem::Discriminant does not implement Ord...
+            let discriminant = |i: &CanonicalIdentity<T>| match i.0 {
+                Identity::Polynomial(..) => 0,
+                Identity::Lookup(..) => 1,
+                Identity::Permutation(..) => 2,
+                Identity::Connect(..) => 3,
+            };
+
+            discriminant(self)
+                .cmp(&discriminant(other))
+                .then_with(|| match (self.0, other.0) {
+                    (
+                        Identity::Polynomial(PolynomialIdentity { expression: a, .. }),
+                        Identity::Polynomial(PolynomialIdentity { expression: b, .. }),
+                    ) => a.cmp(b),
+                    (
+                        Identity::Lookup(LookupIdentity {
+                            left: a, right: b, ..
```

Comment:
You could use a generic `left` and `right` function for any non-polynomial identity.

Reply:
I thought about it but given that these are in different variants, I don't see how we could use it:
now:
```rust
Identity::Lookup(LookupIdentity { left, right, .. }) | Identity::Permutation(PermutationIdentity { left, right, ..}) => {
   left
}
```

With a trait, we can't treat both variants at the same time because we can only bind to variables of the same type in patterns:
```rust
// does NOT work
Identity::Lookup(i) | Identity::Permutation(i) => {
   i.left()
```

For this to be possible, we need to use the same struct inside the variant:

```
Identity::Lookup(ConnectingIdentity<Lookup>),
Identity::Permutation(ConnectingIdentity<Permutation>),
```

... but this doesn't work either because the generic types also need to match!

So we go one step further and use the same variant

```
Identity::Connection(ConnectingIdentity)
```

And store the kind inside the `ConnectingIdentity`.

This is nice but my issue is that at this point we drifted away completely from the PIL enum.

My conclusion is that relying on patterns to work across connecting identities, while verbose, is fine as it preserves the structure of the PIL enum. If we refactor the PIL enum to be nested, then the nested approach would work in Rust as well. I don't think this refactor is necessarily beneficial (because of its nested nature), nor a priority.